// @mintlify/validation/dist/openapi/convertServers.js
export var convertServers = function (_a) {
    var servers = _a.servers;
    if (servers === undefined || servers.length === 0) {
        return undefined;
    }
    return servers.map(function (_a) {
        var url = _a.url, description = _a.description, variables = _a.variables;
        return {
            url: url,
            description: description,
            variables: convertVariables({ variables: variables }),
        };
    });
};
var convertVariables = function (_a) {
    var variables = _a.variables;
    if (variables === undefined || Object.keys(variables).length === 0) {
        return undefined;
    }
    var newEntries = Object.entries(variables).map(function (_a) {
        var name = _a[0], variable = _a[1];
        if (variable.enum) {
            return [
                name,
                {
                    type: 'stringEnum',
                    enum: variable.enum,
                    description: variable.description,
                    default: variable.default,
                },
            ];
        }
        return [
            name,
            {
                type: 'string',
                description: variable.description,
                default: variable.default,
            },
        ];
    });
    return Object.fromEntries(newEntries);
};
