// @mintlify/cli/src/helpers.ts
import { getConfigPath } from '@mintlify/prebuild';
import detect from 'detect-port';
import inquirer from 'inquirer';
import Ora, { Ora as OraType } from 'ora';
import { createInterface } from 'readline';
import shell from 'shelljs';
import { ArgumentsCamelCase } from 'yargs';

import { CMD_EXEC_PATH } from './constants.js';

const confirm = async (question: string) =>
  new Promise((resolve) => {
    const cmdInterface = createInterface({
      input: process.stdin,
      output: process.stdout,
    });
    cmdInterface.question(question, (response) => {
      cmdInterface.close();
      resolve(response.toLowerCase() === 'y');
    });
  });

export const checkPortRecursive: <T>(
  argv: ArgumentsCamelCase<T>
) => Promise<number | undefined> = async (argv) => {
  const port = (argv.port || 3000) as Parameters<typeof detect>[0];
  const _port = await detect(port);

  if (port == _port) {
    return port;
  }

  const confirmed = await confirm(
    `Port ${port} is already in use. Use port ${_port} instead? [Y/n]\n`
  );

  if (confirmed) {
    return await checkPortRecursive({
      ...argv,
      port: _port,
    });
  }
  return undefined;
};

export const promptForYarn = async () => {
  const yarnInstalled = shell.which('yarn');
  if (!yarnInstalled) {
    await inquirer
      .prompt([
        {
          type: 'confirm',
          name: 'confirm',
          message: 'yarn must be globally installed. Install yarn?',
          default: true,
        },
      ])
      .then(({ confirm }) => {
        if (confirm) {
          shell.exec('npm install --global yarn');
        } else {
          console.log('Installation cancelled.');
        }
      });
  }
};

export const buildLogger = (startText = ''): OraType => {
  const logger = Ora().start(startText);
  return logger;
};

export const checkForMintJson = async () => {
  if (await getConfigPath(CMD_EXEC_PATH)) return;
  console.error('Must be run in a directory where a mint.json file exists.');
  process.exit(1);
};
